[FEAT] hub integration with dataset and configuration

burtenshaw opened the pull request:
This is an experimental WIP pr to get feedback on the approach. I've migrated the code out of the v1 argilla client, and reimplemented for v2 changes. It would be great to get feedback on issues like this:

- testing with a mocked hub api.
- dealing with responses across argilla servers and mismatched user_ids.
- dealing with dependencies like `huggingface_hub`. I like the decorator used in the v1 client.

Here's a dataset that I pushed. Still uses default readme from v1:

https://huggingface.co/datasets/burtenshaw/test-argilla-dataset

To test this implementation do:

```python
import uuid

from datetime import datetime

import argilla as rg

client = rg.Argilla(api_key="owner.apikey")
workspace = client.workspaces[0]
mock_dataset_name = (
    f"test_add_record_with_suggestions {datetime.now().strftime('%Y%m%d%H%M%S')}"
)
mock_data = [
    {
        "text": "Hello World, how are you?",
        "label": "positive",
        "id": uuid.uuid4(),
        "comment": "I'm doing great, thank you!",
        "topics": ["topic1", "topic2"],
        "topics.score": [0.9, 0.8],
    },
    {
        "text": "Hello World, how are you?",
        "label": "negative",
        "id": uuid.uuid4(),
        "comment": "I'm doing great, thank you!",
        "topics": ["topic3"],
        "topics.score": [0.9],
    },
    {
        "text": "Hello World, how are you?",
        "label": "positive",
        "id": uuid.uuid4(),
        "comment": "I'm doing great, thank you!",
        "comment_score": 0.9,  # This field will be ignored because it is not in the mapping
        "rating": 1,
        "topics": ["topic1", "topic2", "topic3"],
        "topics.score": [0.9, 0.8, 0.7],
        "ranking": ["label1", "label2", "label3"],
        "span": [
            {
                "start": 0,
                "end": 5,
                "label": "label1",
            },
            {
                "start": 6,
                "end": 11,
                "label": "label2",
            },
            {
                "start": 12,
                "end": 17,
                "label": "label3",
            },
        ],
        "vector": [1, 2, 3],
    },
]
settings = rg.Settings(
    fields=[
        rg.TextField(name="text"),
    ],
    questions=[
        rg.LabelQuestion(name="label", labels=["positive", "negative"]),
        rg.RatingQuestion(name="rating", values=[1, 2, 3, 4, 5]),
        rg.RankingQuestion(name="ranking", values=["label1", "label2", "label3"]),
        rg.TextQuestion(name="comment", use_markdown=False),
        rg.MultiLabelQuestion(
            name="topics",
            labels=["topic1", "topic2", "topic3"],
            labels_order="suggestion",
        ),
        rg.SpanQuestion(
            name="span", labels=["label1", "label2", "label3"], field="text"
        ),
    ],
    metadata=[
        rg.FloatMetadataProperty(name="comment_score"),
    ],
    vectors=[
        rg.VectorField(name="vector", dimensions=3),
    ],
)
dataset = rg.Dataset(
    name=mock_dataset_name,
    settings=settings,
    client=client,
)
dataset.create()
dataset.records.log(
    mock_data,
    mapping={
        "comment": "comment.suggestion",
        "comment_score": "comment.suggestion.score",  # This field will be ignored because it is not in the mapping
        "topics": "topics.suggestion",
        "topics.score": "topics.suggestion.score",
        "label": "label.response",
        "span": "span",
    },
)

dataset.to_hub(repo_id="burtenshaw/test-argilla-dataset")


pulled_dataset = from_huggingface("burtenshaw/test-argilla-dataset")
```

burtenshaw committed:
test: update tests for refactored mapping method

burtenshaw committed:
refactor: introduce independent mapping method and move logic to before ingestion loop

burtenshaw committed:
docs: update all doc strings in dataset records

burtenshaw committed:
chore: improve typing and docs on type

burtenshaw committed:
docs: wrong method in records api reference

burtenshaw committed:
feat: add exception for record ingestion

burtenshaw committed:
refactor: improve explainabilitity and readability in ingestion code with refactor and tqdm and exceptions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
enhancement: move mapping out of record loop

frascuchon committed:
[REFACTOR] Avoid autofetch when accessing settings (#5112)

Reviewing and improving records.log

Instead of: 

<img width="1335" alt="Captura de pantalla 2024-06-26 a las 12 48 14"
src="https://github.com/argilla-io/argilla/assets/2518789/02283f4c-fe6a-464f-96b3-36853e6c7622">

for 50 records, records.log can log 1000:

<img width="870" alt="Captura de pantalla 2024-06-26 a las 12 48 57"
src="https://github.com/argilla-io/argilla/assets/2518789/d20f0469-0b33-427e-aa12-b4b7e1d40cd1">

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

burtenshaw committed:
enhancement: use just one progress bar

burtenshaw committed:
chore: update typing of mapping

burtenshaw committed:
fix: move render mapping into infer record method

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: align add records parameters with render function

burtenshaw committed:
feat: implement ingestion mapping as class

burtenshaw committed:
feat: use ingestion mapping class in dataset records not dataset records

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
chore: tidy imports

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

burtenshaw committed:
docs: update mapping parameters in how to guides

burtenshaw committed:
test: broaden suggestion mapping in test

burtenshaw committed:
feat: add flat for including records in export

burtenshaw committed:
feat: extract dot notation with regex not string splitting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
docs: typo in docs

burtenshaw committed:
feat: improve record switch in ingest method

burtenshaw committed:
feat: refactor id mapping away from dict to type

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
feat: use class methods for type and parameter values

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

frascuchon committed:
[REFACTOR] generate default mapping and extends it with custom mapping dict (#5151)

<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

Closes #<issue_number>

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Refactor (change restructuring the codebase without changing
functionality)

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- follows the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

burtenshaw committed:
refactor: migrate mapper into module from file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: raise error not warn when mapped attribute is unknown

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' of https://github.com/argilla-io/argilla into spike/mapping-to-tuple

frascuchon committed:
Merge branch 'develop' into spike/mapping-to-tuple

burtenshaw committed:
Merge branch 'spike/mapping-to-tuple' into feat/hub-integration

burtenshaw committed:
feat: migrate disk method into module not file

burtenshaw committed:
feat: migrate default card out of argilla v1

burtenshaw committed:
feat: migrate huggingface hub methods from v1 into sdk

burtenshaw committed:
feat: inherit export to hub mixin in dataset resource class

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
Merge remote-tracking branch 'origin/develop' into feat/hub-integration

burtenshaw committed:
Merge branch 'feat/hub-integration' of https://github.com/argilla-io/argilla into feat/hub-integration

davidberenstein1957 commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,182 @@
+            size_categories_parser,
+        )
+
+        hfds = self.records(with_vectors=True, with_responses=True, with_suggestions=True).to_datasets()
```
perhaps we should `with_*` as kwargs? Thinking about that, should we add a `with_metadata`?

davidberenstein1957 commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,182 @@
+            _default_settings_path = os.path.join(tmpdirname, self._default_settings_path)
+            _default_dataset_path = os.path.join(tmpdirname, self._default_dataset_path)
+
+            HfApi().upload_file(
```
should we initialize the `HfApi` once at the beginning of the code¿

frascuchon replied:
Maybe is easier to upload all the dataset json configuration under a remote  `.argilla` folder. With this, we can just upload the temporary folder using the `HfApi().upload_folder` method. Also, in the future, we can store under the `.argilla` folder the SDK version used for this to maintain these export features.

davidberenstein1957 commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,182 @@
+
+        hfds = load_dataset(repo_id)
+
+        if isinstance(hfds, DatasetDict) and "split" not in kwargs:
```
I have been thinking about this before too. Because splits are quite fundamental in ML and the Hugging Face Hub, perhaps we can consider it as a reserved metadata field in future versions.

davidberenstein1957 commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,182 @@
+            records_with_responses.append(record)
+        dataset.records.log(records_with_responses)
+
+        dataset.records.log(hfds)
```
```suggestion
```

davidberenstein1957 commented on argilla/src/argilla/datasets/_export/card/_parser.py:
```diff
@@ -0,0 +1,35 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+AVAILABLE_SIZE_CATEGORIES = {
```
maybe we can import this from the Hugging Face hub library?

davidberenstein1957 replied:
We discussed adding this as lightweight core dependency because of telemetry too.

Wauplin replied:
this doesn't exist in `huggingface_hub` but could be added if relevant. Do you know if `datasets` has a similar mapping?

davidberenstein1957 commented on argilla/src/argilla/datasets/_export/card/__init__.py:
```diff
@@ -0,0 +1,20 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from ._dataset_card import ArgillaDatasetCard
```
maybe use full imports?

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,182 @@
+            hfds = hfds[list(hfds.keys())[0]]
+
+        mapping = {col: col for col in hfds.column_names if ".suggestion" in col}
+        dataset.records.log(hfds, mapping=mapping)
```
Does it make sense to support the mapping value as an argument and keep this as the default one?

frascuchon replied:
Also, this default mapping is not already covered in the `IngestedRecordMapper` implementation?

burtenshaw replied:
Correct `IngestedRecordMapper` does not have a persistent default mapping. I think this is a separate PR though?

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/_hub.py

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
refactor: simplify code using folder not files and split up logic

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,169 @@
+                argilla_guidelines=self.settings.guidelines or None,
+                argilla_vectors_settings=self.settings.vectors or None,
+                argilla_metadata_properties=self.settings.metadata,
+                argilla_record=sample_argilla_record.api_model().model_dump_json(),
```
Why not using `record.to_dict()`? This dump is an internal record state and users may be confused.

burtenshaw committed:
feat: assign missing users to me user

burtenshaw committed:
feat: add with records param

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,190 @@
+            hf_dataset: Dataset = hf_dataset[list(hf_dataset.keys())[0]]
+ 
+        if with_records:
+            dataset.create()
```
Maybe users expect the dataset will be created even if they don't want to add records. And also this can create weird flows:
```python
ds = Dataset.from_hub(..., with_records=False)
ds.exists() # will be False/None

ds = Dataset.from_hub(..., with_records=True)
ds.exists() # will be True
```
What about defining an explicit flag to create the dataset (which can be automatically enabled when with_records is True)?

frascuchon commented on argilla/src/argilla/datasets/_export/_disk.py:
```diff
@@ -39,7 +39,7 @@ class DiskImportExportMixin(ABC):
     _default_dataset_path = "dataset.json"
 
-    def to_disk(self: "Dataset", path: str) -> str:
+    def to_disk(self: "Dataset", path: str, include_records: bool = True) -> str:
```
we can align this `include_records` with the new `with_records` since both have similar purposes.

burtenshaw committed:
feat: reuse hf dataset dependency by migrating to helpers

burtenshaw committed:
fix: client usage in response user_id lookup

burtenshaw committed:
Merge branch 'feat/hub-integration' of https://github.com/argilla-io/argilla into feat/hub-integration

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

dvsrepo commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+            **kwargs: the kwargs to pass to `datasets.Dataset.load_from_hub`.
+
+        Returns:
+            A `FeedbackDataset` loaded from the Hugging Face Hub.
```
```suggestion
            A `Dataset` loaded from the Hugging Face Hub.
```

dvsrepo commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+        *args: Any,
+        **kwargs: Any,
+    ):
+        """Loads a `FeedbackDataset` from the Hugging Face Hub.
```
```suggestion
        """Loads a `Dataset` from the Hugging Face Hub.
```

dvsrepo commented on argilla/src/argilla/datasets/_export/card/argilla_template.md:
```diff
@@ -0,0 +1,210 @@
+---
```
All this will need to be updated before the release

burtenshaw replied:
Yep. I'm on that today.

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/_hub.py

Co-authored-by: Daniel Vila Suero <daniel@argilla.io>

burtenshaw committed:
chore: make with_records and include_records consistent

burtenshaw committed:
enhancement: use to_dict method

burtenshaw committed:
docs: refine default template for dataset card

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
chore: use full imports in card module

dvsrepo commented on argilla/src/argilla/datasets/_export/card/argilla_template.md:
```diff
@@ -0,0 +1,185 @@
+- **Point of Contact:** {{ point_of_contact }}
+{% endif %}
+
+This dataset has been created with [Argilla](https://argilla-io.github.io). As shown in the sections below, this dataset can be loaded into Argilla as explained in [Load with Argilla](#load-with-argilla), or used directly with the `datasets` library in [Load with `datasets`](#load-with-datasets).
```
```suggestion
This dataset has been created with [Argilla](https://argilla-io.github.io). As shown in the sections below, this dataset can be loaded into Argilla as explained in [Using this dataset with Argilla](#using-this-dataset-with-argilla), or used directly with the `datasets` library in [Using this dataset with datasets](#using-this-dataset-with-datasets).
```

dvsrepo replied:
The link doesn't seem to work, should we use the github link as an entry point (there's an overview, link to docs, and we might even get a star?)

dvsrepo replied:
```suggestion
This dataset has been created with [Argilla](https://argilla-io.github.io). As shown in the sections below, this dataset can be loaded into your Argilla server as explained in [Load with Argilla](#load-with-argilla), or used directly with the `datasets` library in [Load with `datasets`](#load-with-datasets).
```

dvsrepo commented on argilla/src/argilla/datasets/_export/card/argilla_template.md:
````diff
@@ -0,0 +1,185 @@
+ds = rg.Dataset.from_hub("{{ repo_id }}")
+```
+
+This will load the settings and records from the dataset repository and push them to the Argilla instance.
````
```suggestion
This will load the settings and records from the dataset repository and push them to you Argilla server for exploration and annotation.
```

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+    from argilla import Argilla
+    from argilla import Workspace
+
+HFDataset = resolve_hf_datasets_type()
```
For what I've seen, `HFDataset` is used only for typing. Is there a reason why not directly doing:

```py
if TYPE_CHECKING:
    from argilla import Dataset
    from argilla import Argilla
    from argilla import Workspace
    from datasets import Dataset as HFDataset
```
?

Since all the code inside `if TYPE_CHECKING` is never executed, you won't have an import issue even if `datasets` is not installed. And this way is more explicit than `resolve_hf_datasets_type` in my opinion (+ it avoids an import at runtime if `datasets` is not needed)

burtenshaw replied:
We were avoiding having `datasets` as a dependency for the core `argilla` install. Instead, it's installed under a parameter `argilla[io]`. But I also question whether that still a good idea.

@frascuchon What do you think? Should we make datasets a core dependency and simplify this?

Wauplin replied:
To be confirmed but having `datasets` under `if TYPE_CHECKING` statement shouldn't make it required since it'll never be executed.

frascuchon replied:
They are different topics. +1 to define hf Dataset under the `TYPE_CHECKING` section (thanks @Wauplin). Even if we define datasets as core dep, we can improve runtime for those scenarios where `datasets` is not required.  We also need to adapt import [here](https://github.com/argilla-io/argilla/blob/feat/hub-integration/argilla/src/argilla/records/_dataset_records.py#L26), with this, we don't need anymore the line:

```python
HFDataset = resolve_hf_datasets_type()
```

About the core dep. What about the `huggingface_hub` dependency? This dep is a transitive dependency of datasets, right? Maybe including them as part of the core dependencies makes more sense now. Otherwise, we should review other pieces of code importing those packages.

Wauplin replied:
Just to confirm, `huggingface_hub` is required by `datasets` but not the other way around.

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+    _default_dataset_repo_dir_name = ".argilla"
+    _default_configuration_files = ["settings.json", "dataset.json"]
+
+    def to_hub(
```
Usually this method is called `push_to_hub`. Feel free to ignore if there is good reason for `to_hub`.

burtenshaw replied:
This is a good question. Our reason is to make all export method's consistent as `to_*`.

Wauplin replied:
It's already a good enough reason!

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+                path_in_repo=self._default_dataset_repo_dir_name,
+            )
+
+        if generate_card:
```
To reduce the number of commits, you can move the `if generate_card:` section under `with TemporaryDirectory() as tmpdirname:`. Then, instead of `card.push_to_hub(repo_id, repo_type="dataset", token=kwargs.get("token"))`, you can do `card.save(os.path.join(tmpdirname, "README.md"))`. And then you move the `hf_api.upload_folder` call at the end to upload both the dataset and the dataset card in a single commit.

Wauplin replied:
**EDIT:** to do the above, you would need to update `self.to_disk` as well:
```py
self.to_disk(path=os.path.join(tmpdirname, self._default_dataset_repo_dir_name), with_records=False)
```

so that the whole `tmpdirname` folder can be uploaded at the root of the repo on the Hub:

```py
hf_api.upload_folder(folder_path=tmpdirname, repo_id=repo_id, repo_type="dataset")
```

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+            size_categories_parser,
+        )
+
+        hf_api = HfApi()
```
better practice to set the `token` once and for all in `HfApi` definition

```suggestion
        hf_api = HfApi(token=kwargs.get("token"))
```

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+
+        with TemporaryDirectory() as tmpdirname:
+            for filename in cls._default_configuration_files:
+                hf_hub_download(
```
Better not to do this in a temporary directory. Using `local_dir` will download it do a temporary location. If you don't use this, the cache directory will be used, meaning that next time the files will already be downloaded. So you could do:

```py
from huggingface_hub import snapshot_download

folder_path = snapshot_download( # download both files in parallel
    repo_id=repo_id,
    repo_type="dataset",
    allow_patterns=cls._default_configuration_files
)

dataset = cls.from_disk(path=folder_path, target_workspace=workspace, client=client)
```

Wauplin replied:
Maybe use `kwargs.get("token")` btw? (to be consistent with `to_hub`)

Wauplin commented on argilla/src/argilla/datasets/_export/card/__init__.py:
```diff
@@ -0,0 +1,20 @@
+#  Copyright 2021-present, the Recognai S.L. team.
```
```suggestion
#  Copyright 2024-present, Argilla, Inc.
```
(nit, to fix in some other files as well)

Wauplin commented on argilla/src/argilla/datasets/_export/card/_dataset_card.py:
```diff
@@ -0,0 +1,30 @@
+    `argilla/client/feedback/integrations/huggingface/card/argilla_template.md`.
+    """
+
+    default_template_path = TEMPLATE_ARGILLA_DATASET_CARD_PATH
```
:+1:

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+
+class HubImportExportMixin:
+    _default_dataset_repo_dir_name = ".argilla"
+    _default_configuration_files = ["settings.json", "dataset.json"]
```
Are those private attributes supposed to change over time? If not, I would make them uppercase (`_DEFAULT_DATASET_REPO_DIR_NAME`) to make it more explicit

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+HFDataset = resolve_hf_datasets_type()
+
+
+class HubImportExportMixin:
```
For what I've read, I feel that this mixin assume that `DiskImportExportMixin` has been injected to the core class as well since `to_hub` is using `to_disk` and `from_hub` is using `from_disk`.

This means `HubImportExportMixin` is dependent on `DiskImportExportMixin`. Don't know if there is a good way of ensuring this in Python except making `HubImportExportMixin` depend on `DiskImportExportMixin`

frascuchon replied:
Yes. I think this definition would clarify dependency on `DiskImportExportMixin`
```suggestion
class HubImportExportMixin(DiskImportExportMixin):
```

Wauplin reviewed (commented):
Hey there :wave: I've made a quick review of the Hub integration after a ping on Slack. I've left a couple of comments on things that could be optimized (I think). Feel free to ignore some comments if irrelevant in the Argilla context. In general, the integration is already quite good as is! Love the auto-generated Argilla dataset card :heart:

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/_hub.py

Co-authored-by: Lucain <lucainp@gmail.com>

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/card/__init__.py

Co-authored-by: Lucain <lucainp@gmail.com>

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/card/argilla_template.md

Co-authored-by: Daniel Vila Suero <daniel@argilla.io>

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/card/argilla_template.md

Co-authored-by: Daniel Vila Suero <daniel@argilla.io>

burtenshaw committed:
docs: update card prose

frascuchon commented on argilla/src/argilla/datasets/_export/card/_dataset_card.py:
```diff
@@ -0,0 +1,30 @@
+
+from pathlib import Path
+
+from huggingface_hub import DatasetCard
```
we definitely need this as part of the core deps.

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+
+        if with_records:
+            cls._extract_suggestions(hf_dataset=hf_dataset, dataset=dataset)
+            cls._extract_responses(hf_dataset=hf_dataset, dataset=dataset)
```
Is there a way to send suggestions and responses in the same .log call? Otherwise, we're processing the HF dataset, and sending the same records twice.

burtenshaw replied:
I've implemented this using `IngestedRecordMapper` directly. I know this is a bit weird, but it means that we can use the mapping and still apply logic to all `user_id`'s when we init `Record` classes.

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+                dataset._client.users._api.get(user_id)
+            except NotFoundError:
+                warnings.warn(message=f"User {user_id} not found. Assigning responses to current user..")
+                user_ids[user_id] = dataset._client.me.id
```
This record will fail when sent to the server if more than one user is not valid.

burtenshaw replied:
Sorry. I don't get this comment. 

My intention is to replace all user_ids in the dataset's responses with the my own user id, if the user_id does not exist on the server.

frascuchon replied:
My comment is that we only cannot provide multiple user responses for the same questions. If more than one user is not found, the same user ID will be assigned to multiple responses, which will raise an error when sending the records to the server.

frascuchon replied:
What about just discarding responses when the user is not found, at least for now?

burtenshaw replied:
Thanks for clarifying. Yes, I guess that's the best option. There are two more details to the implementation:

- discard the responses after the first from unknown `user_id`'s, so the first response is still ingested
- discard all responses from unknown user_id's and warn with the unknown user_id's

I guess all of these would be best dealt with via a feature in the Mapper to map response users.

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,194 @@
+                    response = Response(
+                        user_id=user_ids[user_id],
+                        question_name=question_name,
+                        value=value,
```
I realize that we should map the status related to the whole user response. I will create a new PR to export the response status as we do for response users.

frascuchon replied:
https://github.com/argilla-io/argilla/pull/5200

burtenshaw replied:
Approved

burtenshaw committed:
fix: link in default readme

burtenshaw committed:
feat: move datasets under type checking

burtenshaw committed:
chore: make constants upper case

burtenshaw committed:
feat: use snapshot download

burtenshaw committed:
chore: use token in from_hub

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
feat: reduce commits to hf repo by uploading folder with card

burtenshaw committed:
Merge branch 'feat/hub-integration' of https://github.com/argilla-io/argilla into feat/hub-integration

burtenshaw committed:
chore: add huggingface hub to dependencies

burtenshaw committed:
feat: use ingested record mapper to avoid double push

burtenshaw committed:
Merge branch 'develop' into feat/hub-integration

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,188 @@
+                folder_path=tmpdirname,
+                repo_id=repo_id,
+                repo_type="dataset",
+                token=kwargs.get("token"),
```
do we need this argument since the hf_api is already created with the same token?

burtenshaw replied:
I've noticed that it's a common pattern. I think it's to deal with multiple token role assignments, for example, across organisations. 

@Wauplin You recommended making token consistent, but would you also recommend exposing it as a param like this?

Wauplin replied:
There are two topics here:
1. Passing `token=kwargs.get("token")` in `hf_api.upload_folder` is indeed not needed since `hf_api` is instantiated as `hf_api = HfApi(token=kwargs.get("token"))`.
2. Is it necessary to expose `token` as a parameter? => this depends on how Argilla is used. In general the preferred way to authenticate is to set the `HF_TOKEN` environment variable or to run `huggingface-cli login` in a terminal which will save your token in a file on the machine. This is the preferred way since it insures users won't paste their token in the code directly.
_However_, it only works if the script/tool perform all actions as the same user. If the tool is running on a server and multiple users can access it and perform actions with their own token, then it's useful to have a `token` parameter. For example, a lot of Gradio Spaces have a "Sign in with Hugging Face" button to retrieve a user token. In the process, the Space might need to save an artifact under the user namespace on HF (say a dataset) => it has to use the `token` parameter.

burtenshaw replied:
Thanks @Wauplin . I'm fixing point 1. On 2, I do think it's useful to expose the token in this way, at least via kwargs.

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,188 @@
+                dataset._client.users._api.get(user_id)
+            except NotFoundError:
+                warnings.warn(message=f"User {user_id} not found. Assigning responses to current user..")
+                user_ids[user_id] = dataset._client.me.id
```
You can use the exists method instead and check with an if-condition
```suggestion
            
            if not dataset._client.users._api.exists(user_id):
                warnings.warn(message=f"User {user_id} not found. Assigning responses to current user..")
                user_ids[user_id] = dataset._client.me.id
```

frascuchon committed:
[BUGFIX] Add response status on record export (#5200)

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/_hub.py

Co-authored-by: Paco Aranda <francis@argilla.io>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

Wauplin commented on argilla/pyproject.toml:
```diff
@@ -13,6 +13,7 @@ dynamic = ["version"]
 dependencies = [
     "httpx>=0.26.0",
     "pydantic>=2.6.0, <3.0.0",
+    "huggingface_hub>=0.0.12",
```
```suggestion
    "huggingface_hub>=0.22.0",
```

`0.0.12` is a veeeery old release :smile:. Latest is 0.23.4 (soon 0.24.x). I would advice to require at least `0.22.0` which includes quite a lot of bug fixes.

burtenshaw committed:
fix: lose token from `upload_folder` method

Wauplin commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -12,29 +12,15 @@
-
-
-HFDataset = _resolve_hf_datasets_type()
+HFDataset = resolve_hf_datasets_type()
```
(nit) Since `resolve_hf_datasets_type` is only used in this file in the end, I feel that it would make it more readable to have the conditional import only in this module (one less module to maintain).

Also I think that doing `try: import ... except ImportError: return None` in `resolve_hf_datasets_type` hides the main `ImportError` exception. Below, there is a `dataset_class = resolve_hf_datasets_type() / if dataset_class is None: / raise ImportError(...)`. Something like this:

```py
    try:
        from datasets import Dataset
    except ImportError as e:
        raise ImportError("Hugging Face datasets is not installed. Please install it using `pip install datasets`.") from e
``` 

would provide more information to the end user in case the `datasets` import fails. At some point in `huggingface_hub` we had a similar pattern and hiding the initial import gave us an headache because the `import datasets` was not failing because `datasets` was uninstalled but because a dependency of `datasets` was not correctly installed (it was not with `datasets` specifically but that's the idea).

burtenshaw replied:
Thanks @Wauplin . `resolve_hf_datasets_type`'s location in `_helpers` is an artefact from sharing the function with the `hub_integration` methods.

I think we should consider making `datasets` a core dependancy because it's use has spread. @frascuchon What do you think?

Wauplin replied:
> resolve_hf_datasets_type's location in _helpers is an artefact from sharing the function with the hub_integration methods.

I think that as per https://github.com/argilla-io/argilla/pull/5161#discussion_r1672263458, this is not the case anymore (`resolve_hf_datasets_type's` seems to be only used in `argilla/src/argilla/records/_io/_datasets.py` as before this PR)

burtenshaw committed:
chore: bump huggingface_hub version to 0.22.0

burtenshaw committed:
feat: adjust logic to deal with multiple responses on a record

burtenshaw committed:
chore: make dataset a core dependency

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: deprecate _hf_datasets helper submodule due to dependency change

burtenshaw committed:
Merge branch 'feat/hub-integration' of https://github.com/argilla-io/argilla into feat/hub-integration

burtenshaw committed:
Merge branch 'feat/hub-integration' of https://github.com/argilla-io/argilla into feat/hub-integration

burtenshaw committed:
Merge branch 'feat/hub-integration' of https://github.com/argilla-io/argilla into feat/hub-integration

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,199 @@
+                        value=value,
+                        status=status,
+                    )
+                    record.responses.add(response)
```
```suggestion
                    if user_id:
                       response = Response(
                           user_id=user_id,
                           question_name=question_name,
                           value=value,
                           status=status,
                       )
                       record.responses.add(response)
```

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,199 @@
+
+        # Create a mapper to map the Hugging Face dataset to a Record object
+        mapping = {col: col for col in hf_dataset.column_names if ".suggestion" in col}
+        mapper = IngestedRecordMapper(dataset=dataset, mapping=mapping, user_id=dataset._client.me.id)
```
Could we remove the `user_id` argument here? Users are already processed in the next section, right?

burtenshaw replied:
user_id is a required parameter of the ingestion mapper.

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,199 @@
+                response_status = values["status"][idx]
+                for value, user_id, status in zip(response_values, response_users, response_status):
+                    user_id = _user_ids.pop(UUID(user_id), None)
+                    if user_id is None:
```
Here, we can have an error if the record contains responses from two unknown users. Those will be mapped to the default user and the add will raise an error.

frascuchon replied:
I think this change allows what you want: If a user is already used (is in responses_users), the response is skipped.
```suggestion
        for idx, row in enumerate(hf_dataset):
            record = mapper(row)
            record.id = row.pop("id")
            for question_name, values in response_questions.items():
                response_users = {}
                response_values = values["responses"][idx]
                response_users = values["users"][idx]
                response_status = values["status"][idx]
                for value, user_id, status in zip(response_values, response_users, response_status):
                    user_id = user_ids[UUID(user_id)]
                    if user_id is in response_users:
                        continue
                    response_users[user_id] = True   
```

frascuchon commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,199 @@
+                response_questions[question_name]["status"] = hf_dataset[col]
+
+        # Check if all user ids are known to this Argilla client
+        known_users_ids = [user.id for user in dataset._client.users._api.list()]
```
using public client api?
```suggestion
        known_users_ids = [user.id for user in dataset._client.users]
```

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/_hub.py

Co-authored-by: Paco Aranda <francis@argilla.io>

burtenshaw committed:
Update argilla/src/argilla/datasets/_export/_hub.py

Co-authored-by: Paco Aranda <francis@argilla.io>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
enhance: make me user a variable

burtenshaw committed:
fix: gihub suggestion syntax error

burtenshaw committed:
chore: clarify inheritance in dataset resource import export mixins

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,202 @@
+        generate_card: Optional[bool] = True,
+        *args,
+        **kwargs,
+    ) -> None:
```
```suggestion
    def to_hub(
        self: "Dataset",
        repo_id: str,
        *,
        generate_card: Optional[bool] = True,
        **kwargs,
    ) -> None:
```

(nit) Personal preference but I would not accept `*args` as parameter + I would force to use a keyword-argument for `generate_card`. My reasoning is that positional args are harder to maintain if you want to rename, deprecate or move them. Keyword arguments on the other hand are easier to maintain.

Also accepting `*args` to forward to `hfds.push_to_hub` do not provide additional features (kwargs is enough) but could harm maintenance if you want to add a new positional argument in the future.

Wauplin commented on argilla/src/argilla/datasets/_export/_hub.py:
```diff
@@ -0,0 +1,202 @@
+        with_records: bool = True,
+        *args: Any,
+        **kwargs: Any,
+    ):
```
(same comment as above)

Wauplin reviewed (commented):
Final review from my side. Integration looks good to me! :heavy_check_mark: 
Left 2 small comments regarding future maintenance.

burtenshaw committed:
[DOCS] add guide for exporting datasets and or their records (#5197)

These PR introduces a new guide on exporting datasets and/or their
records. It moves content out of the current 'query and export' guide,
and creates a soley export guide.

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Daniel Vila Suero <daniel@argilla.io>

burtenshaw committed:
Merge branch 'develop' into feat/hub-integration

burtenshaw committed:
fix: remove old liscence

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
chore: lose args in favour of * for the sake of maintenance

burtenshaw committed:
chore: lose args

burtenshaw committed:
Merge remote-tracking branch 'origin/develop' into feat/hub-integration

burtenshaw committed:
fix: remove all old liscences

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci
